File: insertSort.py
def insertSort(nums):
    n = len(nums)
    for i in range(1, n):
        for j in range(0, i):
            if nums[i] <= nums[j]:   # 如果第i个元素比前面第j个元素小，则将它移至第j个元素处，后面的元素依次向后移动一个位置
                temp = nums[i]
                for k in range(i, j, -1):
                    nums[k] = nums[k-1]   # 依次向后移动一个位置
                nums[j] = temp
    return nums

a = [125, 130, 132, 123, 127, 129, 117, 121, 126, 116, 120, 122]
print(insertSort(a))

File: insertSort.go
package main

import (
	"fmt"
)

func InsertSort(nums []int) int {
	// 复杂度为O(N^2)，最坏情况为(N-1)N/2，最好情况为O(N)，即遍历N-1次
	count := 0
	for p := 1; p < len(nums); p++ {
		i := 0
		temp := nums[p]
		for i = p; i > 0; i-- {
			count++ // 需要比较的次数，从后往前插入，遇到比插入元素小的就停止遍历内层循环
			if nums[i-1] <= temp {
				break
			}
			nums[i] = nums[i-1]
			// count++            需要交换的次数，等于逆序对数
		}
		nums[i] = temp
	}
	return count
}

func main() {
	a := [6]int{34, 8, 64, 51, 32, 21}
	count := InsertSort(a[:])
	fmt.Println(a)
	fmt.Println(count)
}
